featurebyte/query_graph/node/function.py

```diff
@@ -0,0 +1,227 @@
+"""
+This module contains generic function related node classes
+"""
+# DO NOT include "from __future__ import annotations" as it will trigger issue for pydantic model nested definition
+from typing import Any, Dict, List, Literal, Optional, Sequence, Tuple, Union
+from typing_extensions import Annotated
+
+from abc import abstractmethod  # pylint: disable=wrong-import-order
+
+from pydantic import BaseModel, Field
+
+from featurebyte.enum import DBVarType, FunctionParameterInputForm
+from featurebyte.models.base import PydanticObjectId
+from featurebyte.query_graph.enum import NodeOutputType, NodeType
+from featurebyte.query_graph.node.base import BaseSeriesOutputNode
+from featurebyte.query_graph.node.metadata.operation import (
+    DerivedDataColumn,
+    NodeOutputCategory,
+    OperationStructure,
+    OperationStructureBranchState,
+    OperationStructureInfo,
+    PostAggregationColumn,
+)
+from featurebyte.query_graph.node.metadata.sdk_code import (
+    CodeGenerationConfig,
+    CodeGenerationContext,
+    ExpressionStr,
+    InfoDict,
+    RightHandSide,
+    StatementT,
+    ValueStr,
+    VariableNameGenerator,
+    VarNameExpressionInfo,
+)
+
+
+class BaseFunctionParameterInput(BaseModel):
+    """BaseFunctionParameterInput class"""
+
+    dtype: DBVarType
+
+    @abstractmethod
+    def get_column_args(self) -> List[Optional[str]]:
+        """
+        Get column arguments (column name or None) for the function parameter input.
+        If the input is a value, return empty list.
+
+        Returns
+        _______
+        List[Optional[str]]
+        """
+
+    @abstractmethod
+    def get_sdk_function_argument(
+        self, node_inputs: List[VarNameExpressionInfo], node_input_index: int
+    ) -> Tuple[RightHandSide, int]:
+        """
+        Get SDK function argument for the function parameter input.
+
+        Parameters
+        ----------
+        node_inputs : List[VarNameExpressionInfo]
+            List of node inputs.
+        node_input_index : int
+            The index of the node input that the function parameter input is associated with.
+
+        Returns
+        -------
+        Tuple[RightHandSide, int]
+            Tuple of SDK function argument and the updated node input index.
+        """
+
+
+class ValueFunctionParameterInput(BaseFunctionParameterInput):
+    """ValueFunctionParameterInput class"""
+
+    value: Optional[Any]
```

Comment:
Would the type that we use for scalar parameters be appropriate here? https://github.com/featurebyte/featurebyte/blob/8377eb48f17d77f1f6cd5637b18074ec08a915bb/featurebyte/query_graph/node/scalar.py#L59

Seems better if we can avoid Any if not necessary.

Reply:
Will update this to use `ValueParameterType`